codegen/gce/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+use ir::{
+    ast::constants::ConstantType::*,
+    transition_stmts::{ConstantValue, Operation},
+    AirIR, BoundaryExpr,
+};
+
+mod boundary_constraints;
+use boundary_constraints::set_boundary_expressions_and_outputs;
+
+mod error;
+use error::ConstraintEvaluationError;
+
+mod helpers;
+use helpers::{accumulate_constants, count_boundary_rand_values, Expression, NodeType};
+
+mod transition_constraints;
+use transition_constraints::{set_transition_expressions, set_transition_outputs};
+
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::Write;
+
+/// Holds data for JSON generation
+#[derive(Default, Debug)]
+pub struct CodeGenerator {
+    num_polys: u16,
+    num_variables: usize,
+    constants: Vec<u64>,
+    expressions: Vec<Expression>,
+    outputs: Vec<usize>,
+}
+
+impl CodeGenerator {
+    pub fn new(ir: &AirIR, extension_degree: u8) -> Result<Self, ConstraintEvaluationError> {
+        // vector of all boundary constraints vectors
+        let boundary_constraints_vec = [
+            &ir.main_first_boundary_constraints(),
+            &ir.main_last_boundary_constraints(),
+            &ir.aux_first_boundary_constraints(),
+            &ir.aux_last_boundary_constraints(),
+        ];
+
+        // maps indexes in Node vector in AlgebraicGraph and in `expressions` JSON array
+        let mut expressions_map = BTreeMap::new();
+
+        // maps names of named constants and public inputs to their NodeType
+        // Only CONST or VAR allowed
+        let mut const_public_type_map = BTreeMap::new();
+
+        // vector of expression nodes
+        let mut expressions = Vec::new();
+
+        // vector of `expressions` indexes
+        let mut outputs = Vec::new();
+
+        let num_polys = set_num_polys(ir, extension_degree);
+        let num_variables =
+            set_num_variables(ir, &mut const_public_type_map, boundary_constraints_vec);
+
+        // TODO #1: get rid of the vector and push values directly into result string
+        // constants from Constants AirIR field
+        // TODO #2: currently I add all found constants in the vector. Should I add only unique ones,
+        // since I'll get constants by their value, not index?
+        let constants = set_constants(ir, &mut const_public_type_map, boundary_constraints_vec);
```

Comment:
I'm not sure if I understood this question correctly, but we should be referencing constants by index, so yes, it would be best to only add unique ones.

In other words, if we had `a'= a + 1`, we would have something like the following:

```
...
        "constants": [1]
	"expressions": [
		{"op":"ADD", "lhs": {"type": "POL", "index": 0}, "rhs": {"type": "CONST", "index": 0}},
		{"op": "SUB", "lhs": {"type": "POL_NEXT", "index": 0}, "rhs": {"type": "REF", "index": 0}}
	],
	"outputs": [
		{"type": "POL_NEXT", "index": 0},
		{"type": "POL", "index": 0},
		{"type": "CONST", "index": 0},
		{"type": "REF", "index": 0}
	]
```